keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+is an extension of DRA that addresses this by making claim parameters less
+opaque. "Semantic models" define how resource capacity per node can be published
+and the claim parameters that allocate some of that capacity for a claim.
+
+This KEP describes the core changes needed for this. Specific semantic models
+get defined in separate KEPs. The types defined by these models become
+part of the resource.k8s.io API. Therefore developers who want to extend
+a scheduler with some additional model have to modify the apiserver.
+
+When opting into using semantic parameters for their DRA driver, driver
+developers no longer need to provide a control plane controller which
+handles claim allocation. Instead, they need to provide a CRD controller
+which translates from their custom claim parameters CRD into the parameter
+object understood by Kubernetes.
+
+Scheduling performance is expected to become better compared to using the
+PodSchedulingContext. However, supporting CA is the main reason for this KEP.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without semantic parameters. Ideally, users should not notice
+  at all that a driver is using semantic parameters under the hood.
+
+- Keep the core DRA functionality from [KEP
+  #3063](https://github.com/kubernetes/enhancements/issues/3063) working as
+  defined there for Kubernetes 1.29. Cluster autoscaling is not always needed
+  and in-tree numeric model(s) are not expected to cover all use case, so the
+  flexibility offered by opaque parameters remains relevant.
+
+### Non-Goals
+
+- Support semantic class parameters. They remain opaque and can be used to
+  define configuration of resources, but must not affect where a claim can be
+  allocated.
+
+- Define exactly one semantic model that covers all current and future resources.
+  There cannot be a "one size fits all" solution.
+
+## Proposal
+
+Each node publishes information about the nearly-opaque resources it currently
+has. Each driver manages one or more resource instance per node. An instance
+could be a single discreet piece of hardware like a PCI card or a pool of such
```

Comment:
> When would I lump 4 GPU cards into 1 "instance" vs. 4 instances? Is an instance an addressable device (like a PCIe BDF)? Or is it a card which has multiple functions?

I think this is getting to the bottom of the problem we are facing right now. It could be both.

When I started, I wanted to define that inside *the* (yes, only one) semantic model that describes an instance. I said "instance" is a bad term because that model could also describe a pool of items. When it is a pool, there may be constraints on how different items may get combined together. All of that would have been encapsulated inside the description of that one semantic model. Because they would have been versioned separately, they could have evolved as needed.

Then John suggested that we turn that "inside out" and make different semantic models composeable, with very simple semantic models. This only works if we move semantic up into the API. For example, suppose a driver manages two discrete cards, and each card has compute units and RAM that can be reserved. This could get described as:

```
kind: NodeResourceCapacity
...
instances:
  - name: CardA
    labels: [...]
    attributes:
    - name: RAM
      counter:
        count: 1Gi
    - name: ComputeUnits
      counter:
        count: 100
  - name: CardB
    labels: [...]
    attributes:
    - name: RAM
      counter:
        count: 20Gi
   -  name: ComputeUnits
      counter:
         count: 500
```

When someone asks for 1Gi RAM, 100 compute units, the semantic would be "find **one** instance that has enough of both" (attributes ANDed together) and an allocation refers to that one instance by name.

When treating GPUs as anonymous items that are part of the pool, then we could have:

```
kind: NodeResourceCapacity
...
instances:
  - name: BigCards
    labels: [...]
    attributes:
    - name: NumberOfCards
      counter:
        count: 1
  - name: SmallCards
    labels: [...]
    - name: NumberOfCards
      counter:
        count: 1 
```

Now when someone asks for "NumberOfCards 2", this could be satisfied by taking cards from *any* instance. This is a different semantic. There's also no notion of sub-dividing an individual card anymore.

Perhaps we can add a field in the claim parameters that says "take from one instance" vs. "from many". At that point we have dropped "semantic models" as a concept and focus entirely on evolving the `resource.k8s.io` API to make it more and more flexible. This worries me.

In the examples above, I already moved `labels` up from being something that is implemented and checked by a "selector" semantic model into the main API. Whether this is a `map[string]string` or something more complex to make CEL expression fully type-safe would need to be discussed as part of the main API.

If we want to do the same for "counter", then we also lift that up:
```
instances:
  - name: CardA
    labels: [...]
    counters:
    - name: RAM
      count: 1Gi
    - name: ComputeUnits
      count: 100
```

In the Go API, that means removing all "one-of-many" structs like `InstanceAttributes`. If we add new fields, they'll be silently ignored when the scheduler doesn't know about them because there won't be a way for it to detect that it got nothing where it should have had something. This probably goes to far.

Reply:
I would be nice if the Claim could say: I want multiple units of a device, but they all should have the same label X (where this label represents the topology).